content/blog/angular-extend-class/index.md

```diff
@@ -0,0 +1,842 @@
+  }
+
+  height = this.window.innerHeight;
+  width = this.window.innerWidth;
+
+  // This needs to be an arrow function
+  onResize = () => {
+    this.height = this.window.innerHeight;
+    this.width = this.window.innerWidth;
+  }
+
+  ngOnInit() {
+    this.window.addEventListener('resize', this.onResize);
+  }
+
+  ngOnDestroy() {
+    this.window.removeEventListener('resize', this.onResize);
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>The window is {{height}}px high and {{width}}px wide</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit, OnDestroy {
+}
+```
+
+
+
+## Overwriting `constructor` behavior {#overwriting-constructors}
+
+When working with class extension, regardless of being used in Angular or in JavaScript itself, you need to call `super()` when trying to overwrite a constructor:
+
+```javascript
+class BaseClass {
+  name = "";
+  constructor() {
+    name = "Frank";
+  }
+}
+
+class AppClass extends BaseClass {
+	constructor() {
+		// This is required
+    super();
+	}
+}
+```
+
+Without calling `super`, you'll get the following error:
+
+```
+Uncaught ReferenceError: must call super constructor before using 'this' in derived class constructor
+```
+
+Likewise, you need to call `super` when overwriting a class component's `constructor` as well.
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent {
+  name = "";
+  constructor() {
+    this.name = "Kevin";
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>{{name}}</p>
+  `,
+})
+class AppComponent extends BaseComponent {
+  constructor() {
+    super();
+    this.name = "Corbin";
+  }
+}
+```
+
+This water gets muddied when using dependency injection in a base component that utilizes dependency injection.
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent {
+  window!: Window;
+  constructor(@Inject(DOCUMENT) private document: Document) {
+    this.window = document.defaultView!;
+  }
+
+  // ...
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent {
+  // This code doesn't work. Read on to learn why
+  constructor() {
+    super();
+  }
+}
+```
+
+As the `super` method needs to be passed with the same arguments from dependenct injection, least we see the following error:
+
+```
+TS2554: Expected 1 arguments, but got 0.
+  app.component.ts(8, 15): An argument for 'document' was not provided.
+```
+
+To solve this, we need to pass `document` from a new instance of `AppComponent`'s dependency injection to `BaseComponent`:
+
+```typescript
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  // This code doesn't work. Read on to learn why
+  constructor(@Inject(DOCUMENT) private document: Document) {
+    super(document);
+    console.log(document.body);
+  }
+}
+```
```

Comment:
As of Angular 14, we have a way to circumvent this: We use the `inject` function in the base class:

```typescript
import { inject } from '@angular/core';

@Component({
  template: ''
})
abstract class BaseComponent {
  // The `Document` type annotation is optional as it can be inferred by the `inject` function
  private document: Document = inject(DOCUMENT);
  
  window!: Window = this.document.defaultView;

  // No constructor needed as we use property injection instead of constructor injection
  
  // ...
}
```

which removes the need for a sub-class to inject the dependencies of a base class:

```typescript
@Component({
  selector: 'app-root',
  template: `
    <p>Test</p>
  `,
})
class AppComponent extends BaseComponent {
  // This code now works as the base class doesn't have constructor parameters anymore
  constructor() {
    super();
  }
}
```